### PR TITLE
WIP : CI Signal Policy Violation: A DashboardTab config MUST CONTAIN an alert contact email address.

### LINKED ISSUE #18599: Kubernetes CI Policy: all prow.k8s.io jobs must have testgrid alerts configured
Part of https://github.com/kubernetes/test-infra/issues/18551

**Why this is important:**
- In order to ensure effective use of community resources, we expect them to be spent on jobs that provide useful signal, and are actively maintained
- Configuring testgrid alerts requires setting an e-mail address, which gives us a point of contact to escalate to if a job is deemed an ineffective use of community resources
- We'll use this to implement a policy where we reserve the right to remove/disable jobs that are deemed an ineffective use of resources (e.g. perma-failing for O(weeks)) if the point of contact is unresponsive

TODO:
- come up with test to enforce this policy (logging only)
- come up with report to identify jobs and likely candidate sig owners
  - eg: if a job is on "sig-foo"'s dashboard, it's likely they want to own it
  - who to contact for jobs that aren't sigs?
- send out notice to all sigs, give a deadline of N weeks
- any jobs that don't have e-mail addresses configured should be removed
- flip test to failing once all jobs meet the policy

Other thoughts / notes:
- need to parse both prowjob config and testgrid config for this
  - there are some testgrids not populated by prow, we can probably ignore these?
  - there are some prowjobs that don't have all their testgrid config in annotations

/sig testing

### PR DESCRIPTION
**Pull request by RobertKielty**
Some WIP on CI Signal Alert Policy enforcement 

CI Signal Policy Violation: A DashboardTab config MUST CONTAIN an alert contact email address.

Fixes #18599

### CODE REVIEW

**Comment by RobertKielty** [comment]
@spiffxp @BenTheElder @kubernetes/ci-signal when y'all get a chance can you review this? 

I just fixed up the PR so that it is go fmt'd properly :roll_eyes: apols for my numptyness! 

Can't remember why I marked this as WIP? 

Is this change useful/correct as-is or do we need to do more on this?

Are these tests run as part of a CI Job when a new CI job is pushed up? 
If so which job?
If not then I might need a little help with setting that up so the test are run there. 

The following run (on my feature branch) looks like ...

Test output in config_tests.txt on the below local invocation is available for review [here](https://gist.github.com/RobertKielty/84f1fc3b6a755982e5ed9eb087e53755) 

```
$ bazel test //config/tests/testgrids:go_default_test > config_tests.txt
Loading: 
Loading: 0 packages loaded
Analyzing: target //config/tests/testgrids:go_default_test (0 packages loaded, 0 targets configured)
INFO: Analyzed target //config/tests/testgrids:go_default_test (1 packages loaded, 1230 targets configured).

INFO: Found 1 test target...
[0 / 1] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[1,270 / 1,271] Testing //config/tests/testgrids:go_default_test; 1s linux-sandbox
FAIL: //config/tests/testgrids:go_default_test (see /home/rkielty/.cache/bazel/_bazel_rkielty/47d03368e29206548a86533381c34b95/execroot/io_k8s_test_infra/bazel-out/k8-fastbuild/testlogs/config/tests/testgrids/go_default_test/test.log)
[1,270 / 1,271] 1 / 1 tests, 1 failed; Testing //config/tests/testgrids:go_default_test; 7s linux-sandbox
INFO: From Testing //config/tests/testgrids:go_default_test:
Target //config/tests/testgrids:go_default_test up-to-date:
  bazel-bin/config/tests/testgrids/linux_amd64_race_stripped/go_default_test
INFO: Elapsed time: 13.867s, Critical Path: 8.16s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed, 1 test FAILED, 2 total actions
INFO: Build completed, 1 test FAILED, 2 total actions
```

**Comment by RobertKielty** [comment]
You can see the test in action on this PR of course :)

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/19286/pull-test-infra-bazel/1319631568897052672

File: config/tests/testgrids/config_test.go [hunk 1]
```diff
@@ -503,6 +503,34 @@ func TestKubernetesProwInstanceJobsMustHaveMatchingTestgridEntries(t *testing.T)
 	}
 }
 
+// WIP See policy description on GH Issue #18599
+func TestJobsMustHaveTestGridAlertsSetup(t *testing.T) {
+	for _, dashboard := range cfg.Dashboards {
+		for _, dashboardtab := range dashboard.DashboardTab {
+			if dashboardtab.GetAlertOptions() == nil {
+				t.Errorf("FAIL : %s : CI Signal Policy Violation: A DashboardTab config MUST CONTAIN an alert contact email address.", dashboardtab.GetName())
```
[the remark this answers is not in the document]
Comment (spiffxp) [reply config/tests/testgrids/config_test.go:511]: Dashboard tab names are not unique.  Consider instead logging dashboard/tab (see `TestReleaseBlockingJobsMustHaveTestgridDescriptions` for an example)

File: config/tests/testgrids/config_test.go [hunk 2]
```diff
@@ -503,6 +503,34 @@ func TestKubernetesProwInstanceJobsMustHaveMatchingTestgridEntries(t *testing.T)
...
+
+			}
+		}
+	}
+}
+
+// WIP subtest version - no real advantage ideally I would have liked subtests
+// to register as counted failures. AFAIK no idiomatic golang test way of
+// getting subtest stats.
```
Comment (spiffxp) [inline config/tests/testgrids/config_test.go:520]: I'm not 100% clear on this either, don't have time to dig in.  However I feel like you could gather stats and log them if you'd like ("x / y failing policy")  I also think test-case-per-dashboard means hundreds or thousands of tests, which probably wouldn't make sense unless each test covered _all_ policies.

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some suggestions

**Pull request closed without merging**